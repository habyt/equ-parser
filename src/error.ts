
export class EQUError extends Error {
    constructor(msg?: string) {
        super(msg)
    }
}

export class EQULexingError extends EQUError {
    constructor(msg?: string) {
        super(msg)
    }
}

export class EQUParsingError extends EQUError {
    constructor(msg?: string) {
        super(msg)
    }
}

