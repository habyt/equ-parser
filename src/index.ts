export * from "./equ"

export * from "./error"

export { Token, Item } from "./lexer"
export {
    FilterOperator,
    FilterOperand,
    ParseItem,
    ExpressionOperand,
    ExpressionOperator,
    ExpressionItem,
} from "./parser"
