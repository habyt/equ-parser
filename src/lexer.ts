import assert from "assert"
export type Token =
    | "eof"
    | "error"
    | "path"
    | "or"
    | "and"
    | "filtersStart"
    | "filtersEnd"
    | "bundleStart"
    | "bundleEnd"
    | "string"
    | "number"
    | "filterGt"
    | "filterLt"
    | "filterGte"
    | "filterLte"
    | "filterEq"
    | "filterCt"

const filters: { [k: string]: Token } = {
    eq: "filterEq",
    lt: "filterLt",
    gt: "filterGt",
    lte: "filterLte",
    gte: "filterGte",
    ct: "filterCt"
}

const tokenBundleStart = "("
const tokenBundleEnd = ")"

const tokenOr = "|"
const tokenAnd = ","

const tokenFiltersStart = "["
const tokenFiltersEnd = "]"

const tokenFilterIs = ":"

const tokenValidPath =
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._"
const tokenValidOperator = "abcdefghijklmnopqrstuvwxyz"
const tokenValidNumberStart = "+-.0123456789"

type StateFn = ((lexer: Lexer) => StateFn) | undefined

export class EQUParsingError extends Error {
    constructor(msg?: string) {
        super(msg)
    }
}

function lexError(msg?: string): never {
    throw new EQUParsingError(msg)
}

class Lexer {
    input: string
    pos: number = 0
    items: Array<Item> = []
    start: number = 0
    width: number = 0

    constructor(str: string) {
        this.input = str
    }

    emit(token: Token) {
        let value = this.input.substring(this.start, this.pos)

        if (token === "string") {
            value = value.replace("\\\\", "\\")
            value = value.replace('\\"', '"')
        }

        this.items.push({ str: value, type: token })
        this.start = this.pos
    }

    next(): string {
        const next = this.input.substr(this.pos, 1)
        this.width = next.length
        this.pos += this.width
        return next
    }

    accept(valid: string): boolean {
        if (valid.indexOf(this.next()) >= 0) {
            return true
        }

        this.backup()
        return false
    }

    backup() {
        this.pos -= this.width
    }

    peek(): string {
        const next = this.next()
        this.backup()
        return next
    }

    acceptRun(valid: string) {
        while (true) {
            const next = this.next()
            if (next === "") {
                break
            }

            if (valid.indexOf(next) === -1) {
                break
            }
        }

        this.backup()
    }

    acceptRunUntil(...stops: Array<string>) {
        while (true) {
            const next = this.next()
            if (next === "") {
                this.backup()
                return false
            }

            if (stops.indexOf(next) >= 0) {
                this.backup()
                return true
            }
        }
    }

    ignore() {
        this.start = this.pos
    }

    rest(): string {
        return this.input.substring(this.start)
    }

    current(): string {
        return this.input.substring(this.start, this.pos)
    }
}

export type Item = {
    str: string
    type: Token
}

function lexFilters(l: Lexer): StateFn {
    if (l.peek() === tokenBundleStart) {
        return lexToken(tokenBundleStart, "bundleStart", lexFilters)
    }

    l.acceptRun(tokenValidPath)
    if (l.pos <= l.start) {
        throw new EQUParsingError("expected path, but got " + l.rest())
    }

    l.emit("path")

    if (l.peek() !== tokenFiltersStart) {
        throw new EQUParsingError("expected '[', but got " + l.rest())
    }

    return lexToken(tokenFiltersStart, "filtersStart", lexInsideFilters)
}

function lexInsideFilters(l: Lexer): StateFn {
    if (l.peek() === tokenBundleStart) {
        return lexToken(tokenBundleStart, "bundleStart", lexInsideFilters)
    }

    l.acceptRun(tokenValidOperator)
    const nextFilter = l.current()
    const filterType =
        filters[nextFilter] ??
        lexError("expected filter type, got " + nextFilter)
    l.emit(filterType)

    if (l.next() !== tokenFilterIs) {
        lexError("expected " + tokenFilterIs + " but got " + l.current())
    }

    l.ignore()

    return lexFilterValue
}

function lexFilterValue(l: Lexer): StateFn {
    const next = l.peek()

    if (tokenValidNumberStart.indexOf(next) >= 0) {
        return lexNumberValue
    }

    if (next === '"') {
        l.next()
        l.ignore()
        return lexStringValue
    }

    lexError("expected filter value but got " + l.rest())
}

function lexStringValue(l: Lexer): StateFn {
    while (true) {
        const found = l.acceptRunUntil("\\", '"')
        if (!found) {
            lexError("expected string to end, but got " + l.rest())
        }

        const next = l.peek()
        if (next === '"') {
            l.emit("string")

            l.next()
            l.ignore()
            return lexAfterFilterValue
        }

        // if we're here next must be a \ because of acceptRunUntil
        assert(next === "\\")
        l.next()

        const escapedSymbol = l.peek()

        if (escapedSymbol === "\\" || escapedSymbol === '"') {
            l.next()
        } else {
            lexError("invalid escape sequence: \\" + escapedSymbol)
        }
    }
}

function lexNumberValue(l: Lexer): StateFn {
    l.accept("+-")

    let decimalOccured = l.accept(".")

    l.acceptRun("0123456789")
    if (l.accept(".")) {
        if (decimalOccured) {
            lexError("invalid number: " + l.current())
        }

        l.acceptRun("0123456789")
    }

    l.emit("number")
    return lexAfterFilterValue
}

function lexAfterFilterValue(l: Lexer): StateFn {
    const next = l.next()
    switch (next) {
        case tokenOr: {
            l.emit("or")
            break
        }
        case tokenAnd: {
            l.emit("and")
            break
        }
        case tokenFiltersEnd: {
            l.backup()
            return lexToken(tokenFiltersEnd, "filtersEnd", lexAfterFilter)
        }
        case tokenBundleEnd: {
            l.backup()
            return lexToken(tokenBundleEnd, "bundleEnd", lexAfterFilterValue)
        }
    }

    return lexInsideFilters
}

function lexAfterFilter(l: Lexer): StateFn {
    const next = l.next()

    if (next === "") {
        l.emit("eof")
        return undefined
    }

    switch (next) {
        case tokenOr: {
            l.emit("or")
            return lexFilters
        }
        case tokenAnd: {
            l.emit("and")
            return lexFilters
        }
        case tokenBundleEnd: {
            l.backup()
            return lexToken(tokenBundleEnd, "bundleEnd", lexAfterFilter)
        }
    }

    lexError("invalid combinator: " + next)
}

function lexToken(tokenValue: string, token: Token, next: StateFn): StateFn {
    return (lexer: Lexer) => {
        lexer.pos += tokenValue.length
        lexer.emit(token)
        return next
    }
}

export type LexerOptions = {
    debug?: boolean
}

export function lex(str: string): Array<Item> {
    const lexer = new Lexer(str)

    let state: StateFn = lexFilters
    while (state !== undefined) {
        state = state(lexer)
    }

    return lexer.items
}
