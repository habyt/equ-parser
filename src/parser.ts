import { Item, EQUParsingError } from "./lexer"

function expressionPrecedence(
    a: ExpressionOperator,
    b: ExpressionOperator
): boolean {
    const list = ["or", "and", "bundleStart", "bundleEnd"]

    return list.indexOf(a.operator) > list.indexOf(b.operator)
}

function filterPrecedence(a: FilterOperator, b: FilterOperator): boolean {
    const list = ["or", "and", "bundleStart", "bundleEnd"]

    return list.indexOf(a.operator) > list.indexOf(b.operator)
}

function parseError(msg?: any): never {
    if (msg === undefined) {
        throw new EQUParsingError()
    }

    if (typeof msg === "string") {
        throw new EQUParsingError(msg)
    }

    throw new EQUParsingError("unimplemented " + JSON.stringify(msg))
}

type StateFn = ((ctx: ParserContext) => StateFn) | undefined

type FilterOperatorType = "and" | "or" | "bundleStart" | "bundleEnd"

type ExpressionType = "eq" | "gt" | "gte" | "lt" | "lte" | "ct"
type ExpressionOperatorType = "and" | "or" | "bundleStart" | "bundleEnd"

type ValueType = "number" | "string"

export type FilterOperator = {
    type: "operator"
    operator: FilterOperatorType
}

export type FilterOperand = {
    type: "operand"
    path: string
    expressions: Array<ExpressionItem>
}

export type ParseItem = FilterOperand | FilterOperator

export type ExpressionOperand = {
    type: "expressionOperand"
    expressionType: ExpressionType
    valueType: ValueType
    value: string | number
}

export type ExpressionOperator = {
    type: "expressionOperator"
    operator: ExpressionOperatorType
}

export type ExpressionItem = ExpressionOperator | ExpressionOperand

class ParserContext {
    items: Array<Item>
    parsedItems: Array<ParseItem> = []

    filterOperatorStack: Array<FilterOperator> = []
    expressionOperatorStack: Array<ExpressionOperator> = []

    constructor(items: Array<Item>) {
        this.items = items
    }

    next(): Item {
        return this.items.shift() ?? parseError("expected to have more items")
    }

    pushFilterItem(item: ParseItem) {
        this.parsedItems.push(item)
    }

    pushExpressionItem(item: ExpressionItem) {
        const current =
            this.currentFilterOperand() ??
            parseError("expected to have a filter operand")

        current.expressions.push(item)
    }

    popFilterOperator(): FilterOperator | undefined {
        return this.filterOperatorStack.pop()
    }

    popExpressionOperator(): ExpressionOperator | undefined {
        return this.expressionOperatorStack.pop()
    }

    topExpressionOperator(): ExpressionOperator | undefined {
        if (this.expressionOperatorStack.length === 0) {
            return undefined
        }

        return this.expressionOperatorStack[
            this.expressionOperatorStack.length - 1
        ]
    }

    topFilterOperator(): FilterOperator | undefined {
        if (this.filterOperatorStack.length === 0) {
            return undefined
        }

        return this.filterOperatorStack[this.filterOperatorStack.length - 1]
    }

    currentFilterOperand(): FilterOperand | undefined {
        for (let i = this.parsedItems.length - 1; i >= 0; i--) {
            const item = this.parsedItems[i]
            if (item.type === "operand") {
                return item
            }
        }

        return undefined
    }
}

function parsePath(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "bundleStart") {
        return parseFilterBundleStart
    }

    if (next.type === "path") {
        ctx.pushFilterItem({
            type: "operand",
            path: next.str,
            expressions: []
        })

        return parseFiltersStart
    }

    parseError(next)
}

function parseFilterBundleStart(ctx: ParserContext): StateFn {
    ctx.filterOperatorStack.push({
        type: "operator",
        operator: "bundleStart"
    })

    return parsePath
}

function parseFiltersStart(ctx: ParserContext): StateFn {
    const next = ctx.next()
    if (next.type === "filtersStart") {
        return parseExpression
    }

    parseError(next)
}

function parseExpression(ctx: ParserContext): StateFn {
    const expressionType = ctx.next()

    if (expressionType.type === "bundleStart") {
        return parseExpressionBundleStart
    }

    let expressionOperator: ExpressionType = "eq"
    switch (expressionType.type) {
        case "filterCt": {
            expressionOperator = "ct"
            break
        }
        case "filterEq": {
            expressionOperator = "eq"
            break
        }
        case "filterGt": {
            expressionOperator = "gt"
            break
        }
        case "filterGte": {
            expressionOperator = "gte"
            break
        }
        case "filterLt": {
            expressionOperator = "lt"
            break
        }
        case "filterLte": {
            expressionOperator = "lte"
            break
        }
        default:
            parseError(expressionType)
    }

    const value = ctx.next()
    let valueType: "string" | "number" = "string"
    switch (value.type) {
        case "string": {
            valueType = "string"
            break
        }
        case "number": {
            valueType = "number"
            break
        }
        default:
            parseError(value)
    }

    let parsedValue: number | string = value.str
    if (value.type === "number") {
        parsedValue = Number(value.str)
    }

    ctx.pushExpressionItem({
        type: "expressionOperand",
        expressionType: expressionOperator,
        valueType,
        value: parsedValue
    })

    return parseAfterExpression
}

function parseExpressionBundleStart(ctx: ParserContext): StateFn {
    ctx.expressionOperatorStack.push({
        type: "expressionOperator",
        operator: "bundleStart"
    })

    return parseExpression
}

function parseAfterExpression(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "filtersEnd") {
        const currentFilterOperand = ctx.currentFilterOperand() ?? parseError()
        currentFilterOperand.expressions.push(
            ...ctx.expressionOperatorStack.reverse()
        )
        ctx.expressionOperatorStack = []

        return parseAfterFilter
    }

    if (next.type === "or" || next.type === "and") {
        const expressionOperator: ExpressionOperator = {
            type: "expressionOperator",
            operator: next.type
        }

        let currentTopExpression = ctx.topExpressionOperator()
        while (
            currentTopExpression !== undefined &&
            expressionPrecedence(expressionOperator, currentTopExpression)
        ) {
            const top = ctx.popExpressionOperator() ?? parseError()
            ctx.pushExpressionItem(top)
            currentTopExpression = ctx.topExpressionOperator()
        }

        ctx.expressionOperatorStack.push(expressionOperator)

        return parseExpression
    }

    if (next.type === "bundleEnd") {
        let op = ctx.popExpressionOperator()
        while (op !== undefined) {
            if (op.operator === "bundleStart") {
                break
            }

            ctx.pushExpressionItem(op)
            op = ctx.popExpressionOperator()
        }
        return parseAfterExpression
    }

    parseError(next)
}

function parseAfterFilter(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "eof") {
        ctx.parsedItems.push(...ctx.filterOperatorStack.reverse())
        ctx.filterOperatorStack = []

        return undefined
    }

    if (next.type === "and" || next.type === "or") {
        const filterOperator: FilterOperator = {
            type: "operator",
            operator: next.type
        }

        let currentTopFilter = ctx.topFilterOperator()
        while (
            currentTopFilter !== undefined &&
            filterPrecedence(filterOperator, currentTopFilter)
        ) {
            const top = ctx.popFilterOperator() ?? parseError()
            ctx.pushFilterItem(top)
            currentTopFilter = ctx.topFilterOperator()
        }

        ctx.filterOperatorStack.push(filterOperator)

        return parsePath
    }

    if (next.type === "bundleEnd") {
        let op = ctx.popFilterOperator()
        while (op !== undefined) {
            if (op.operator === "bundleStart") {
                break
            }

            ctx.pushFilterItem(op)
            op = ctx.popFilterOperator()
        }
        return parseAfterFilter
    }

    parseError(next)
}

export function parse(items: Array<Item>): Array<ParseItem> {
    const parser = new ParserContext(items)
    let state: StateFn = parsePath
    while (state !== undefined) {
        state = state(parser)
    }

    return parser.parsedItems
}
