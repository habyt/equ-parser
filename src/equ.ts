import { ParseItem } from "./parser"
import { lex } from "./lexer"
import { parse as parseLex } from "./parser"

export function parse(equ: string): Array<ParseItem> {
    const lexed = lex(equ)

    return parseLex(lexed)
}
