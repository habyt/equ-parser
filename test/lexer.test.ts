import { lex } from "../src/lexer"

const example = `name.first[eq:"foo"|eq:"bar"]|email[ct:"foo",ct:"bar"],age[(gt:4.5,lt:-10)|eq:15]`

describe("lexer", () => {
    it("should lex the example correctly", () => {
        const result = lex(example)

        expect(result).toMatchSnapshot()
    })

    it("should throw an EQUParsingError on error", () => {
        expect(() => lex("ioi[")).toThrowErrorMatchingSnapshot()
    })

    it("should throw a parsing error on unended string values", () => {
        expect(() => lex(`path[eq:"asd`)).toThrowErrorMatchingSnapshot()
    })

    it("should lex booleans correctly", () => {
        const result = lex(`path[ex:true],path2[ex:false]`)

        expect(result).toMatchSnapshot()
    })

    it("should lex bundles correctly", () => {
        const result = lex(`(path[eq:"asd"],path2[eq:"bdf"])`)

        expect(result).toMatchSnapshot()
    })

    it("should lex expression bundles correctly", () => {
        const result = lex(`path[eq:"asd"|gt:"asdf",lt:"asdf"]`)

        expect(result).toMatchSnapshot()
    })

    it("should throw an error if the path is missing", () => {
        expect(() => lex("[")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on missing start filters", () => {
        expect(() => lex("as")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on missing filterIs", () => {
        expect(() => lex("as[eq]")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on an invalid filter value", () => {
        expect(() => lex("as[eq:[]")).toThrowErrorMatchingSnapshot()
    })

    it("should correctly parse esacpe sequences", () => {
        const result = lex(`path[eq:"string with \\\\ in it and also \\""]`)

        expect(result).toMatchSnapshot()
    })

    it("should throw an error on an invalid escape sequence", () => {
        expect(() => lex(`path[eq:"\\s"]`)).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on an invalid number", () => {
        expect(() => lex("path[eq:.12.12]")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on invalid combinator", () => {
        expect(() =>
            lex("path[eq:123]:path[eq:123]")
        ).toThrowErrorMatchingSnapshot()
    })

    it("should lex decimal numbers correctly", () => {
        const result = lex("path[eq:-123.123]")
        const result2 = lex("path[eq:.123]")

        expect(result).toMatchSnapshot()
        expect(result2).toMatchSnapshot()
    })

    it("should lex a date correctly", () => {
        const result = lex("path[eq:2020-01-01]")

        expect(result).toMatchSnapshot()
    })

    it("should lex zulu date time", () => {
        const result = lex("path[eq:2020-01-01T12:00:00.000Z]")

        expect(result).toMatchSnapshot()
    })

    it("should lex offset date time", () => {
        const result = lex("path[eq:2020-01-01T12:00:00.000+01:00]")

        expect(result).toMatchSnapshot()
    })

    it("should not lex dates with invalid formats", () => {
        expect(() => lex("path[eq:2020-1-1]")).toThrowErrorMatchingSnapshot()
    })

    it("should not lex incomplete dates", () => {
        expect(() => lex("path[eq:2020-23]")).toThrowErrorMatchingSnapshot()
    })

    it("should not lex incomplete date time values", () => {
        expect(() =>
            lex("path[eq:2020-10-10T12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123+]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123+12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123+12:]")
        ).toThrowErrorMatchingSnapshot()
    })
})
