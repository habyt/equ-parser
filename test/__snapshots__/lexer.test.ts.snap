// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer should correctly parse esacpe sequences 1`] = `
Array [
  Object {
    "str": "path",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "string with \\\\ in it and also \\"",
    "type": "string",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": "",
    "type": "eof",
  },
]
`;

exports[`lexer should lex booleans correctly 1`] = `
Array [
  Object {
    "str": "path",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "ex",
    "type": "filterEx",
  },
  Object {
    "str": "true",
    "type": "boolean",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": ",",
    "type": "and",
  },
  Object {
    "str": "path2",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "ex",
    "type": "filterEx",
  },
  Object {
    "str": "false",
    "type": "boolean",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": "",
    "type": "eof",
  },
]
`;

exports[`lexer should lex bundles correctly 1`] = `
Array [
  Object {
    "str": "(",
    "type": "bundleStart",
  },
  Object {
    "str": "path",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "asd",
    "type": "string",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": ",",
    "type": "and",
  },
  Object {
    "str": "path2",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "bdf",
    "type": "string",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": ")",
    "type": "bundleEnd",
  },
  Object {
    "str": "",
    "type": "eof",
  },
]
`;

exports[`lexer should lex expression bundles correctly 1`] = `
Array [
  Object {
    "str": "path",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "asd",
    "type": "string",
  },
  Object {
    "str": "|",
    "type": "or",
  },
  Object {
    "str": "gt",
    "type": "filterGt",
  },
  Object {
    "str": "asdf",
    "type": "string",
  },
  Object {
    "str": ",",
    "type": "and",
  },
  Object {
    "str": "lt",
    "type": "filterLt",
  },
  Object {
    "str": "asdf",
    "type": "string",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": "",
    "type": "eof",
  },
]
`;

exports[`lexer should lex the example correctly 1`] = `
Array [
  Object {
    "str": "name.first",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "foo",
    "type": "string",
  },
  Object {
    "str": "|",
    "type": "or",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "bar",
    "type": "string",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": "|",
    "type": "or",
  },
  Object {
    "str": "email",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "ct",
    "type": "filterCt",
  },
  Object {
    "str": "foo",
    "type": "string",
  },
  Object {
    "str": ",",
    "type": "and",
  },
  Object {
    "str": "ct",
    "type": "filterCt",
  },
  Object {
    "str": "bar",
    "type": "string",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": ",",
    "type": "and",
  },
  Object {
    "str": "age",
    "type": "path",
  },
  Object {
    "str": "[",
    "type": "filtersStart",
  },
  Object {
    "str": "(",
    "type": "bundleStart",
  },
  Object {
    "str": "gt",
    "type": "filterGt",
  },
  Object {
    "str": "4.5",
    "type": "number",
  },
  Object {
    "str": ",",
    "type": "and",
  },
  Object {
    "str": "lt",
    "type": "filterLt",
  },
  Object {
    "str": "-10",
    "type": "number",
  },
  Object {
    "str": ")",
    "type": "bundleEnd",
  },
  Object {
    "str": "|",
    "type": "or",
  },
  Object {
    "str": "eq",
    "type": "filterEq",
  },
  Object {
    "str": "15",
    "type": "number",
  },
  Object {
    "str": "]",
    "type": "filtersEnd",
  },
  Object {
    "str": "",
    "type": "eof",
  },
]
`;

exports[`lexer should throw a parsing error on unended string values 1`] = `"expected string to end, but got asd"`;

exports[`lexer should throw an EQUParsingError on error 1`] = `"expected filter type, got "`;

exports[`lexer should throw an error if the path is missing 1`] = `"expected path, but got ["`;

exports[`lexer should throw an error on an invalid escape sequence 1`] = `"invalid escape sequence: \\\\s"`;

exports[`lexer should throw an error on an invalid filter value 1`] = `"expected filter value but got []"`;

exports[`lexer should throw an error on an invalid number 1`] = `"invalid number: .12."`;

exports[`lexer should throw an error on invalid combinator 1`] = `"invalid combinator: :"`;

exports[`lexer should throw an error on missing filterIs 1`] = `"expected : but got ]"`;

exports[`lexer should throw an error on missing start filters 1`] = `"expected '[', but got "`;
